### PR TITLE
Initial implementation of the SRE Challenge

### PR DESCRIPTION
**Pull request by michaelmcallister**
Initial pass of the SRE challenge documentation.

Still left outstanding is the creation of the GitHub repo that we'll use as a seed for this challenge.

Co-authored-by: Philippe M. Chiasson <gozer@goteleport.com>
Signed-off-by: Michael McAllister <michael.mcallister@goteleport.com>

### CODE REVIEW

**Comment by gozer** [comment]
> Still left outstanding is the creation of the GitHub repo that we'll use as a seed for this challenge.

Any clue as to what the repo could be named, I can get it created.

**Comment by jimbishopp** [comment]
Added a commit that duplicates requirements for each level and merges sub-bullets. I'd also like to add a design phase to each level (working on that now). Then I'll add some comments to the PR.

File: challenges/cloud/challenge.md [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+* HTTP API to set the replica count of the Kubernetes Deployment
+
+### Automation
+
+* Ability to build the server in a Docker container
+* Ability to run the server
+* Ability to execute tests against a locally configured Kubernetes config
```
Comment (jimbishopp) [inline challenges/cloud/challenge.md:128]: I wasn't sure what this meant.

File: challenges/cloud/challenge.md [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+* Deploy and manage a local Kubernetes Cluster
+* Deploy and upgrade to the local Kubernetes Cluster with no service interruption
+* GitHub Action driven workflows for Deployment
+
+### Deployment
+
+* Create a helm chart for the service
+* Deploy releases of the API Server to a Kubernetes cluster (and possible dependencies)
+* Upgrade releases of the API Server to a Kubernetes cluster (and possible dependencies)
+* Include a production Deployment of this service: Service, Ingress, TLC, RBAC
```
Comment (jimbishopp) [inline challenges/cloud/challenge.md:221]: Thinking this should be w/L4 as well. We may also want to be more specific about each of the items. Not sure ingress should be included, and RBAC could be interpreted in many different ways.

**Review by jimbishopp — COMMENTED** [review]
I added some cleanup commits and have comments below. Overall I think we're pretty close. My concerns: 1) are the bullet points self-evident?, 2) Is there too much work for L4/5?

File: challenges/cloud/challenge.md (hunk 1 above)
Comment (gozer) [reply challenges/cloud/challenge.md:128]: It's certainly somewhat vague (on purpose?) but the intent was to execute some level of testing against a locally running kubernetes cluster, for instance:
 - deploying the server in k8s and a deployment crd
 - Verifying the server can find the deployment
 - Verifying the server can report the correct number of replicas

Something testing, but closer to integration testing needing to deal with a few more moving parts than a simple mocked unit test

File: challenges/cloud/challenge.md [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+* Secure connections with mTLS
+
+### Automation
+
+* Ability to build the server in a Docker container
+* Ability to run the server
+* Ability to execute tests against a locally configured Kubernetes config
+* GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
+* Build releases of the Server
+* Produce production quality Docker images
```
[the remark this answers is not in the document]
Comment (gozer) [reply challenges/cloud/challenge.md:175]: Yeah, I would just merge them as

- Produce production ready releases (binaries and docker image)

Or some such wording.

File: challenges/cloud/challenge.md (hunk 2 above)
Comment (gozer) [reply challenges/cloud/challenge.md:221]: How about:

Include a production Deployment of this service, including (not limited to) these objects:
 - Deployment
 - Role
 - RoleBinding
 - ServiceAccount
 - Service

We can avoid the Ingress by just requiring the Service

File: challenges/cloud/challenge.md [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+to an AWS EKS cluster.
+
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+* HTTP API to set the replica count of a the Kubernetes Deployment
+* HTTP API to get the list of available Deployments in the Kubernetes cluster
+* HTTP health check verifying Kubernetes connectivity
+* Extend the API to support a difference between current and desired state
+* Extend the Server to support the ability to show the differences between current and desired state
```
[the remark this answers is not in the document]
Comment (jimbishopp) [reply challenges/cloud/challenge.md:165]: @gozer What is our expectation for where they will store expected state?

[the remark this answers is not in the document]
Comment (gozer) [reply challenges/cloud/challenge.md:165]: No idea, the point for me was to be vague on purpose and see what they can come up with, on top of my head I can think of many ways it could be done, in-memory, local pod state, k8s resource annotations/status, external DB (redis/mysql/psql/etc) and so on, all with pluses and minuses. Figured it would be interesting to see _how_ such a requirement would be met, really.

File: challenges/cloud/sre.md [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+
+Open a pull request with your design and share a link with the reviewers via Slack.
+After the document is approved by at least 2 reviews, move on to the implementation.
+
+A few notes about the design document:
+
+* We expect the design document to be completed roughly within the first week.
+  This is to ensure you have enough time to work on the implementation.
+* Avoid writing an overly detailed design document. Two to three pages is
+  sufficient.
```
Comment (wadells) [inline challenges/cloud/sre.md:65]:
```suggestion
* Avoid writing an overly detailed design document. 500-1500 words is
  sufficient.
```

Markdown doesn't neatly map to the pages, making 2-3 pages rather ambiguous.  I'm spitballing words (~500 per page) as an alternative length metric, but I'm open to other ideas.

File: challenges/cloud/sre.md [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+checkout the repository and execute the tests themselves.
+
+Do not try to achieve full test coverage. This will take too long. Write enough
+to exercise the different key components and show they are working as
+intended, as well as to demonstrate your approach to automation.
+
+## Deployment
+
+Levels 3+ include a deployment step which should deploy the solution to a local
+Kubernetes cluster. The choice of which Kubernetes cluster is up to you, but
```
Comment (wadells) [inline challenges/cloud/sre.md:92]: > The choice of which Kubernetes cluster is up to you

This appears to conflict with later guidance to use AWS EKS.  If the intention is to have two deploy targets (e.g. minikube/kind _and_ EKS, that could be clarified).

File: challenges/cloud/sre.md [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+Additional external dependencies are acceptable such as Terraform and minikube/k3d,
+but please ensure your Makefile is handles these dependencies.
+
+# Timing
+
+You can split coding over a couple of weekdays or weekends and find time to ask
+questions and receive feedback.
+
+Once you join the Slack channel, you have between 1 to 2 weeks complete the
+challenge depending on the challenge you choose.
```
Comment (wadells) [inline challenges/cloud/sre.md:373]:
```suggestion
Once you join the Slack channel, you have up to 2 weeks to complete the
challenge.
```

There doesn't appear to be a choice of challenges presently.

File: challenges/cloud/sre.md [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+
+# Tools
+
+This task should be implemented in Go and should work on a 64-bit Linux or MacOS
+machine.
+
+It's safe to assume a working Docker environment will be available locally as well.
+
+Additional external dependencies are acceptable such as Terraform and minikube/k3d,
+but please ensure your Makefile is handles these dependencies.
```
Comment (wadells) [inline challenges/cloud/sre.md:365]:
```suggestion
Additional external dependencies are acceptable such as Terraform and minikube/k3d,
but please ensure your Makefile has targets to install these dependencies.
```

File: challenges/cloud/sre-server.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+	// create the clientset
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	// add handlers
+	http.Handle("/healthz", &healthzHandler{clientset: clientset})
+	http.Handle("/metrics", http.NotFoundHandler()) // TODO: add metrics
```
Comment (wadells) [inline challenges/cloud/sre-server.go:47]: Is this metrics `TODO` something outstanding for us or for the candidate?  I don't see any reference to it later in the challenge document.

File: challenges/cloud/sre.md [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+* HTTP API to set the replica count of a the Kubernetes Deployment
+* HTTP API to get the list of available Deployments in the Kubernetes cluster
+* HTTP health check verifying Kubernetes connectivity
+
+### Automation
+
+* Ability to build the server in a Docker container
+* Ability to run the server
+* Ability to execute integration tests against the local Kubernetes cluster
+* GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
```
Comment (wadells) [inline challenges/cloud/sre.md:142]: > all tests pass

Does this refer to the integration tests from earlier? Are there expected to be any other tests? It'd be worth clarifying to `the integration tests pass` since integration tests in CI is a much bigger lift than `go test ./...` unit tests.

**Review by wadells — COMMENTED** [review]
First round of review gave me a couple minor nits/questions.

I don't have a feel for the bigger "is the scope right" question without trying it out.

File: challenges/cloud/sre.md (hunk 6 above)
Comment (jimbishopp) [reply challenges/cloud/sre.md:92]: We are only asking that they include an EKS section in the design.

File: challenges/cloud/sre.md (hunk 5 above)
Comment (zmb3) [reply challenges/cloud/sre.md:65]: Great suggestion Walt.

File: challenges/cloud/sre.md [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+interview Slack channel. Invite interview panel participants as contributors to
+the new repository if you prefer to keep your submission private.
+
+## Design Doc
+
+Start with a brief design document that covers the edge cases and design approach. At
+Teleport, we prefer Markdown for
+[our designs](https://github.com/gravitational/teleport/blob/master/rfd/0000-rfds.md).
+
+Be sure to cover the following in your design:
```
Comment (zmb3) [inline challenges/cloud/sre.md:50]: Reading this top to bottom, and by the time I get here I still have very little context as to what the thing I'm designing is supposed to do. The only information I've been given so far is the initial sentence that mentions API and Kubernetes.

Would it make sense to add a few more sentences about the project here so the reader has this context in mind when reading about what needs to go in the design doc?

File: challenges/cloud/sre.md [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+
+* Developer workflow
```
Comment (zmb3) [inline challenges/cloud/sre.md:52]: What does this mean? Can we be more specific?

File: challenges/cloud/sre.md [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+* Level 3+: Build, Release
+* Level 4+: State management and Delivery
+* Level 5+: Reconciliation, Conflicts, and Automation
+
+Open a pull request with your design and share a link with the reviewers via Slack.
+After the document is approved by at least 2 reviews, move on to the implementation.
```
Comment (zmb3) [inline challenges/cloud/sre.md:58]:
```suggestion
After the document is approved by at least 2 reviewers, move on to the implementation.
```

File: challenges/cloud/sre.md [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+  sufficient.
+* Avoid sending us draft design documents, it is difficult to evaluate what
+  parts are draft and which parts are complete. Instead we encourage asking
+  questions in Slack and sharing a design document that is ready to be
+  reviewed.
+
+## Server & Automation
+
+Once the design document is approved, begin working on adding features to the
+server implementation that includes automation. A key feature of this challenge
```
Comment (zmb3) [inline challenges/cloud/sre.md:74]: Again, "add features that includes automation" doesn't mean much to me. 

So far, we've asked the candidate to design something and then to start adding features to the something but we have given almost no information as to what the something is.

File: challenges/cloud/sre.md [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+
+* Go
+* make
+* Docker
+
+## Level 1
+
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
```
Comment (zmb3) [inline challenges/cloud/sre.md:108]: I think I see what's been confusing me. The header for "requirements" is on line 24, but we don't actually start talking about requirements for the project until this point. Everything up to now was more about guidance for the interview process and not actual requirements for the thing they need to build.

Maybe it's as simple as changing that first heading to something else so.

File: challenges/cloud/sre.md [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+
+### Automation
+
+* Ability to build the server in a Docker container
```
Comment (zmb3) [inline challenges/cloud/sre.md:112]: Does this mean write a Dockerfile for the server, or are we talking in a more general sense of some automation that produces an image? 

If it's the former and we want to see a Dockerfile, let's be explicit.

Also, the wording of this bullet point is a little weird. "build the server in a Docker container" could mean "build a runnable docker image" or it could mean "spin up a docker container that will run the build and produce an artifact that you can run outside of docker."

File: challenges/cloud/sre.md [hunk 17]
```diff
@@ -0,0 +1,376 @@
...
+
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+* HTTP API to set the replica count of a the Kubernetes Deployment
+* HTTP API to get the list of available Deployments in the Kubernetes cluster
+* HTTP health check verifying Kubernetes connectivity
+* Extend the API to support a difference between current and desired state
+* Extend the Server to support the ability to show the differences between current and desired state
+* Secure connections with mTLS
```
Comment (zmb3) [inline challenges/cloud/sre.md:164]: The connection between the server and the k8s apiserver, the connection between the HTTP API and the caller, or both?

File: challenges/cloud/sre.md [hunk 18]
```diff
@@ -0,0 +1,376 @@
...
+
+### Automation
+
+* Ability to build the server in a Docker container
+* Ability to run the server
+* Ability to execute integration tests against the local Kubernetes cluster
+* GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
+* Produce production ready releases (binaries and docker image)
```
Comment (zmb3) [inline challenges/cloud/sre.md:172]: This is pretty subjective and can mean different things to difference candidates depending on their background. Is there any way we can make this more fair by listing the specific things we'd like to see?

File: challenges/cloud/sre.md [hunk 19]
```diff
@@ -0,0 +1,376 @@
...
+
+### Deployment
+
+* Create a helm chart for the service
+* Deploy releases of the API Server to a Kubernetes cluster (and possible dependencies)
+* Upgrade releases of the API Server to a Kubernetes cluster (and possible dependencies)
+* Include a production Deployment of this service, including but not limited to: Deployment, ServiceAccount, Service
```
Comment (zmb3) [inline challenges/cloud/sre.md:179]: This would likely be done via the helm chart mentioned above, right?

If so, I recommend listing the minimal criteria required for the helm chart at Level 3 (maybe it's just a deployment, and then add whatever else it is we care about here at level 4).

File: challenges/cloud/sre.md [hunk 20]
```diff
@@ -0,0 +1,376 @@
...
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+* HTTP API to set the replica count of a the Kubernetes Deployment
+* HTTP API to get the list of available Deployments in the Kubernetes cluster
+* HTTP health check verifying Kubernetes connectivity
+* Extend the API to support a difference between current and desired state
+* Extend the Server to support the ability to show the differences between current and desired state
+* Secure connections with mTLS
+* Replace the HTTP API with gRPC
```
Comment (zmb3) [inline challenges/cloud/sre.md:199]: In our challenges for Core, switching from. HTTP to gRPC was (IIRC) a level 3 item.

For SRE, maybe this makes sense to show that software engineering skills are more important at higher levels, but I also question whether this is just extra busy work that doesn't really differentiate level 4 and 5 candidates in any dimension other than how much time they have to work on it.

File: challenges/cloud/sre.md [hunk 21]
```diff
@@ -0,0 +1,376 @@
...
+The interview team will join the Slack channel. The team consists of the
+engineers who will be working with you. Ask them about the engineering culture,
+work and life balance, or anything else that you would like to learn about
+Teleport.
+
+Before writing the actual code, create a brief design document in markdown in Github
+and share with the team.
+
+This document should cover at least: design approach, trade-offs, scope,
+proposed API, resiliency, scalability and deployment considerations.
```
Comment (zmb3) [inline challenges/cloud/sre.md:233]: This is somewhat repetitive of the original section, and makes the reader look in multiple places, For example:

- we talk about starting with a design document on line 58 and then introduce the design doc again here
- we mention splitting the code into 2-3 PRs on line 31 but then we talk about splitting into PRs again on line 239

File: challenges/cloud/sre.md (hunk 8 above)
Comment (zmb3) [reply challenges/cloud/sre.md:365]: Seems fine to just have the makefile spit out "terraform: command not found" as long as instructions for how to install it are linked, don't you think?

Make is a bit esoteric and writing makefiles to install tools that could easily be installed via brew (for example) doesn't seem the best use of a candidate's time.

Comment (wadells) [reply challenges/cloud/sre.md:365]: That's fair.  This change was motivated more out of a desitre be clear when we say the `Makefile` should "handle" the dependencies.  Im happy with either an installation target or context aware failures as long as it is clear for the candidate what is expected.

File: challenges/cloud/challenge.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (zmb3) [reply challenges/cloud/challenge.md]: If we're going to be vague on purpose, let's make it clear that we're doing so and say something like "storage of state is left as an exercise for the candidate" or something.

There shouldn't be any "tricks" or other areas where we're expecting the candidate to read our minds and fill in the gaps with where we've secretly omitted details.

File: challenges/cloud/sre.md (hunk 20 above)
Comment (michaelmcallister) [reply challenges/cloud/sre.md:199]: > For SRE, maybe this makes sense to show that software engineering skills are more important at higher levels

This was a core part of why this was proposed, but there are other signals that I thought were important:

- securing via mTLS is expected to be more challenging, and can be more brittle compared to a web server
- We don't provide much in the way of a skeleton for gRPC, so it needs to be implemented from the ground up, rather than adapting what is already there
- how the protobuf contract is proposed is an interesting signal to me, it's more nuanced than HTTP, especially making a protobuf schema future proof 

In a nutshell, I'd probably make the distinguish the difference between the two as: An L4 candidate is expected to augment and adapt an existing codebase, whereas an L5 is expected to innovate and challenge the status quo.

With that said, I'm not super firm on getting these signals from gRPC -- though I do think it mirrors our working environment quite nicely

File: challenges/cloud/sre.md (hunk 21 above)
Comment (michaelmcallister) [reply challenges/cloud/sre.md]: I don't disagree that in parts this section doesn't introduce new information, but as long as we are consistent I don't think it's such an issue to reinforce and condense our expectations under the Guidance section, particularly if it's just a single line. 

With that said, I understand the value of brevity, so I've pared down this to focus on the process and less on expectations, PTAL

File: challenges/cloud/sre-server.go [hunk 22]
```diff
@@ -0,0 +1,65 @@
+package main
```
Comment (michaelmcallister) [inline challenges/cloud/sre-server.go:1]: @jimbishopp will candidates fork this repo, and extend it? Or just copy & paste?

File: challenges/cloud/sre.md (hunk 19 above)
Comment (michaelmcallister) [reply challenges/cloud/sre.md]: Yeah great point, thank you. I've condensed so it's much more clearer.

I don't _believe_ there were extra requirements from L3 -> L4 with respects to Helm (@gozer, correct me if I'm wrong) so both requirements are mirrored.

File: challenges/cloud/sre-server.go (hunk 22 above)
Comment (jimbishopp) [reply challenges/cloud/sre-server.go:1]: I changed the link so it points to the go file. Candidates can click the link and download it. Forking isn't necessary.

File: challenges/cloud/sre.md (hunk 16 above)
Comment (michaelmcallister) [reply challenges/cloud/sre.md]: I agree this is worded confusingly. 

The expectation would be simply to write a Dockerfile to build an image for this server, an implicit expectation when reviewing it would be to use a multi-stage build.

I've updated to (hopefully) be more clearer.

File: challenges/cloud/sre.md (hunk 10 above)
Comment (michaelmcallister) [reply challenges/cloud/sre.md:146]: Yes, there is an expectation to write more tests:

> Add a couple of high quality tests that cover happy and unhappy scenarios. Keep
external dependencies as low or clear as possible. Expect the reviewers to
checkout the repository and execute the tests themselves.

This isn't called out as a line item for each of the levels. So I'll add it there.

There's no expectation for a lot of test coverage, (1 or 2 tests should be okay) but expectations on the quality would increase.

File: challenges/cloud/sre.md (hunk 18 above)
Comment (michaelmcallister) [reply challenges/cloud/sre.md:178]: > This is pretty subjective and can mean different things to difference candidates depending on their background.

This is a feature, not a bug

> Is there any way we can make this more fair by listing the specific things we'd like to see?

We could. We could also field those questions in slack, or the candidate can make assumptions. In any case, how the candidate interprets what it means to be production ready is an interesting signal to me. I'd rather keep this ambiguous so I can see what candidates come up with in the given time constraints rather than be explicit

File: challenges/cloud/sre-server.go (hunk 9 above)
Comment (michaelmcallister) [reply challenges/cloud/sre-server.go:47]: @gozer, @jimbishopp  I believe we dropped metrics, and dashboarding as an ask -- is that correct?

If so we might as well remove this from the server. If not, and it's still a requirement we might be pushing the limit on what's reasonable to expect from our candidates.

Comment (gozer) [reply challenges/cloud/sre-server.go:47]: Yes, we did drop metrics and dashboarding as an ask, IIRC, however, I'd say just leave it there as it is, after all, this is supposed to be a half-made POC with cruft possibly in it. Maybe the candidate leaves it, maybe they clean it up and remove it, maybe they implement it, maybe they ask good questions about telemetry?

I don't see a big downside to leaving it there, just one more source of a potential signal IMO.

File: challenges/cloud/sre.md (hunk 12 above)
Comment (gozer) [reply challenges/cloud/sre.md:52]: Would adding a few bullet points like these be sufficient? I agree it's a bit vague:
  * Ease of contributing to the project from a fresh clone
  * Ease of building, running and testing the server
  * Ease of avoiding common mistakes

File: challenges/cloud/sre.md (hunk 20 above)
Comment (gozer) [reply challenges/cloud/sre.md:204]: > In a nutshell, I'd probably make the distinguish the difference between the two as: An L4 candidate is expected to augment and adapt an existing codebase, whereas an L5 is expected to innovate and challenge the status quo.

I think that's describing it very well, exactly.

File: challenges/cloud/sre.md (hunk 10 above)
Comment (wadells) [reply challenges/cloud/sre.md:146]: Thanks for the clarification. Can we make sure to call out integration testing as explicitly in scope or out of scope for github actions?  I'd recommend:

* unit testing & static analysis/linting is in scope for CI
* integration testing is out of scope for CI

This is mostly to keep the challenge lighter.

Comment (gozer) [reply challenges/cloud/sre.md:146]: How about just adding:

* GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
  * Unit testing and static analysis is in scope
  * Integration testing can be omitted/skipped

File: challenges/cloud/sre.md (hunk 18 above)
Comment (zmb3) [reply challenges/cloud/sre.md:178]: I'll leave this up to you all then, but I'm not sure I agree.

Production ready to someone who's coming from Google is going to mean something different to someone who is maybe newer in their career, has a non-traditional background, or hasn't had the opportunity to work in big tech.

At the end of the day, both people are capable of doing the same work, I want to hire them both, not just the one who had the privilege to work somewhere where "production ready" more closely aligned with what our views.

File: challenges/cloud/sre.md (hunk 10 above)
Comment (wadells) [reply challenges/cloud/sre.md:146]: > How about just adding:
> 
>     * GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
>       
>       * Unit testing and static analysis is in scope
>       * Integration testing can be omitted/skipped

Sounds good to me!

Comment (michaelmcallister) [reply challenges/cloud/sre.md:148]: Added - thanks!

File: challenges/cloud/sre.md (hunk 12 above)
Comment (michaelmcallister) [reply challenges/cloud/sre.md:52]: Added the first two as subpoints, thank you!

File: challenges/cloud/sre-server.go (hunk 9 above)
Comment (michaelmcallister) [reply challenges/cloud/sre-server.go]: Yeah, I think an oversight. I don't remember making a conscious decision together to remove it.

With that said, there's plenty in this challenge to help me form an opinion on a candidates skillset, so let's remove this for now.

File: challenges/cloud/sre.md (hunk 15 above)
Comment (pierrebeaucamp) [reply challenges/cloud/sre.md:110]: 👍 I found myself scrolling back up as I thought I missed the requirements.

File: challenges/cloud/sre.md [hunk 23]
```diff
@@ -0,0 +1,384 @@
...
+deployment automation for an EKS target, or create any AWS resources.
+
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+* HTTP API to set the replica count of a the Kubernetes Deployment
+* HTTP API to get the list of available Deployments in the Kubernetes cluster
+* HTTP health check verifying Kubernetes connectivity
+* Extend the API to support a difference between current and desired state
+* Extend the Server to support the ability to show the differences between current and desired state (implementation of state storage for the server is left as a design choice for the candidate)
```
Comment (pierrebeaucamp) [inline challenges/cloud/sre.md:172]: What's the difference between both points?

**Comment by pierrebeaucamp** [comment]
I don't have to add too much, I think this looks good overall. I do think the difference between L1 and L2 is very small, but then again I don't know if we plan on hiring many L1 SREs

File: challenges/cloud/sre.md (hunk 11 above)
Comment (jimbishopp) [reply challenges/cloud/sre.md:50]: I see the issue that Zac is talking about. I'll add an intro section that summarizes the task.

File: challenges/cloud/sre.md [hunk 24]
```diff
@@ -0,0 +1,384 @@
...
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+* HTTP API to set the replica count of the Kubernetes Deployment
+* One or two tests that cover happy and unhappy scenarios
+
+### Automation
+
+* Write a Dockerfile to build an image for the server
+* Ability to run the server
+* Ability to execute integration tests against the local Kubernetes cluster
+* One or two tests that cover happy and unhappy scenarios
```
[the remark this answers is not in the document]
Comment (jimbishopp) [reply challenges/cloud/sre.md:131]: Looks like a dup.

File: challenges/cloud/sre.md (hunk 23 above)
Comment (jimbishopp) [reply challenges/cloud/sre.md:172]: Yeah, I had the same question in the original version. We've updated it since but I agree that there is still too much overlap.

File: challenges/cloud/sre.md (hunk 18 above)
Comment (jimbishopp) [reply challenges/cloud/sre.md:186]: Yeah, I think we're going to field some questions about this. We'll leave it in for now and update once we get a better idea of how to communicate without ambiguity.

**Review by jimbishopp — APPROVED** [review]

**Review by gozer — APPROVED** [review]
Perfect being the enemy of good, this feels more than good enough to me

**Comment by michaelmcallister** [comment]
Go Team! 

Thanks for all the feedback. Expect us to iterate on this as we gather feedback from our candidates
